### PR TITLE
[Distributed] DeleteKey API for c10d TCP Store

### PR DESCRIPTION
**Pull request by osalpekar**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45223 [Distributed] Adding Python tests for the TCPStore getNumKeys and deleteKey
* **#43963 [Distributed] DeleteKey API for c10d TCP Store**
* #43962 [Distributed] getNumKeys API to c10d TCPStore

This PR adds a deleteKey API to the TCP Store, which deletes the specified key from the store. This API will be useful for some applications related to debug logging in ProcessGroupNCCL going forward. This PR also adds some C++ tests for this API and Python tests are added in #45223. We will build on this functionality in the future by implementing this API for FileStore, HashStore, RedisStore, and ZeusStore.

Differential Revision: [D23009117](https://our.internmc.facebook.com/intern/diff/D23009117/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23009117/)!

### CODE REVIEW

File: caffe2/distributed/file_store_handler.cc [hunk 1]
```diff
@@ -127,6 +127,10 @@ int64_t FileStoreHandler::getNumKeys() {
   return 0;
 }
 
+void FileStoreHandler::deleteKey(const std::string& /* unused */) {
+  CHECK(false) << "deleteKey not implemented for FileStoreHandler";
```
Comment (pritamdamania87) [inline caffe2/distributed/file_store_handler.cc:131]: `TORCH_CHECK`?

File: torch/lib/c10d/test/TCPStoreTest.cpp [hunk 2]
```diff
@@ -40,6 +40,10 @@ void testHelper(const std::string& prefix = "") {
     // We expect 5 keys since 3 are added above, 'counter' is added by the
     // helper thread, and the init key to coordinate workers.
     EXPECT_EQ(numKeys, 5);
+
+    serverStore->deleteKey("key0");
+    numKeys = serverStore->getNumKeys();
```
Comment (pritamdamania87) [inline torch/lib/c10d/test/TCPStoreTest.cpp:45]: nit: probably also test `getKey` returns an error.

File: torch/lib/c10d/Store.hpp [hunk 3]
```diff
@@ -30,6 +30,8 @@ class Store {
 
   virtual int64_t add(const std::string& key, int64_t value) = 0;
 
+  virtual void deleteKey(const std::string& key) = 0;
```
Comment (pritamdamania87) [inline torch/lib/c10d/Store.hpp:33]: Usually a delete API returns true/false about whether the key was removed or not. I think this should return a bool.

File: torch/csrc/distributed/c10d/init.cpp [hunk 4]
```diff
@@ -303,6 +307,10 @@ They are used in specifying strategies for reduction collectives, e.g.,
               "add",
```
Comment (pritamdamania87) [inline torch/csrc/distributed/c10d/init.cpp:307]: All of these would need docs for python users. We can do that in a separate PR.

**Review by pritamdamania87 — CHANGES_REQUESTED** [review]
Requesting changes since I think the API should return a bool

File: torch/lib/c10d/test/TCPStoreTest.cpp (hunk 2 above)
Comment (pritamdamania87) [reply torch/lib/c10d/test/TCPStoreTest.cpp:52]: Can we add a test here for calling `getKey()` on a delete key?

**Review by pritamdamania87 — APPROVED** [review]

**Comment by ngimel** [comment]
Reverting, as it is breaking OSS build
```
08:04:53 /var/lib/jenkins/workspace/caffe2/distributed/redis_store_handler.cc: In member function ‘virtual bool caffe2::RedisStoreHandler::deleteKey(const string&)’:
08:04:53 /var/lib/jenkins/workspace/caffe2/distributed/redis_store_handler.cc:86:1: error: no return statement in function returning non-void [-Werror=return-type]
08:04:53  }
08:04:53  ^
08:04:53 cc1plus: some warnings being treated as errors
08:04:53 caffe2/CMakeFiles/torch_cpu.dir/build.make:10370: recipe for target 'caffe2/CMakeFiles/torch_cpu.dir/distributed/redis_store_handler.cc.o' failed
08:04:53 make[2]: *** [caffe2/CMakeFiles/torch_cpu.dir/distributed/redis_store_handler.cc.o] Error 1
08:04:53 make[2]: *** Waiting for unfinished jobs....
08:04:58 CMakeFiles/Makefile2:6884: recipe for target 'caffe2/CMakeFiles/torch_cpu.dir/all' failed
08:04:58 make[1]: *** [caffe2/CMakeFiles/torch_cpu.dir/all] Error 2
08:04:58 Makefile:140: recipe for target 'all' failed
08:04:58 make: *** [all] Error 2
```
https://ci.pytorch.org/jenkins/job/caffe2-builds/job/pytorch-linux-bionic-rocm3.7-py3.6-build/5732//console

**Pull request closed without merging**